Fix lightbox UI disallow editing

artemiomorales opened the pull request:
## What?
This PR fixes an issue wherein the `allowEditing` value of the lightbox was not resulting in a consistent UX. It also addresses other related edge cases.

## Why?
Fixes https://github.com/WordPress/gutenberg/issues/59837
Fixes https://github.com/WordPress/gutenberg/issues/59797

When one disables editing of the lightbox, the expectation is that the UI related to the lightbox config should be seamlessly removed from the editor.

## How?
It adds logic to handle the case wherein the lightbox animation is enabled by default but editing of the lightbox is not.
It also covers various edge cases regarding how the block-level settings interact with the `theme.json` settings.

## Testing Instructions
To make sure this is working properly, We need to test various scenarios and edge cases.
In particular, we need to test the UI in the following scenarios:

1. Image block DOES NOT have a link configured, and...
- DOES NOT have a lightbox override
- DOES have a lightbox override of ENABLED set to TRUE
- DOES have a lightbox override of ENABLED set to FALSE
2. Image block DOES have a link configured, and...
- DOES NOT have a lightbox override
- DOES have a lightbox override of ENABLED set to TRUE
- DOES have a lightbox override of ENABLED set to FALSE

In addition, each one of the scenarios above needs to be tested with the following `theme.json` configurations:

**allowEditing _false_, enabled _false_**
```
{
	"$schema": "https://schemas.wp.org/trunk/theme.json",
	"version": 2,
	"settings": {
		"blocks": {
			"core/image": {
				"lightbox": {
					"allowEditing": false,
					"enabled": false
				}
			}
		}
	}
}
```

**allowEditing _true_, enabled _false_**
```
{
	"$schema": "https://schemas.wp.org/trunk/theme.json",
	"version": 2,
	"settings": {
		"blocks": {
			"core/image": {
				"lightbox": {
					"allowEditing": true,
					"enabled": false
				}
			}
		}
	}
}
```

**allowEditing _false_, enabled _true_**
```
{
	"$schema": "https://schemas.wp.org/trunk/theme.json",
	"version": 2,
	"settings": {
		"blocks": {
			"core/image": {
				"lightbox": {
					"allowEditing": false,
					"enabled": true
				}
			}
		}
	}
}
```

**allowEditing _true_, enabled _true_**
```
{
	"$schema": "https://schemas.wp.org/trunk/theme.json",
	"version": 2,
	"settings": {
		"blocks": {
			"core/image": {
				"lightbox": {
					"allowEditing": true,
					"enabled": true
				}
			}
		}
	}
}
```

To test:
1. Add the image blocks detailed above to a new post
2. Click on each image, and verify that the UI appears (or is hidden) as expected for each scenario
3. Do this with each specified variation of the `theme.json`

### Testing overview

https://github.com/WordPress/gutenberg/assets/5360536/2dbf9a42-86f4-42a0-9827-9cfa7db1df19

### Test cases walkthrough

#### Part 1
*I mispeak early in this video — in the first scenario, the lightbox UI should be disabled.*

https://github.com/WordPress/gutenberg/assets/5360536/984d907f-7aff-4e20-8066-c2e2b1d4ced9

#### Part 2
https://github.com/WordPress/gutenberg/assets/5360536/09d70ca7-14b5-4237-9d29-a4cdbc4bb7bf

artemiomorales committed:
Check that lightbox can be edited before rendering lightbox UI

artemiomorales commented:
What should the behavior be in the event that the image block has `lightbox.enabled` set to `true`, as in the following example?

```
<!-- wp:image {"lightbox":{"enabled":true|false},"id":30,"sizeSlug":"large","linkDestination":"none"} -->
<figure class="wp-block-image size-large"><img src="http://gutenberglocal.com/wp-content/uploads/2023/12/32259440831_a39a7fb46e_o-1024x768.jpg" alt="" class="wp-image-30"/></figure>
<!-- /wp:image -->
```

This can happen if a user has a local override set to enable or disable the lightbox because of using a `theme,json` that has `lightbox.allowEditing` set to `true`, then later changing to a `theme.json` wherein `lightbox.allowEditing` is `false`.

In this scenario, the block's local lightbox config would supersede the theme.json's default config, but with the UI hidden, it would also be impossible to reset the local value without the opening the code editor or otherwise modifying the block attributes manually.

artemiomorales commented:
I *believe* this PR is working as intended, but will give another test with fresh eyes tomorrow. In any case, I think it's ready to help open up discussion on how to handle block-level overrides as mentioned above.

artemiomorales committed:
Refactor to reduce duplicate code

artemiomorales commented:
Ok this doesn't work as expected yet and mistakenly shows a blank popover if a link is configured and the lightbox's `allowEditing` and `enabled` are both `true`. Am looking into this.

michalczaplinski commented:
> What should the behavior be in the event that an image block has an override of lightbox.enabled

In the ideal world, we would have some kind of "log" and the latest setting would take priority (in your case, the value from the `theme.json`) but such a mechanism does not exist in Gutenberg.

IMO, the least suprising behavior is that the local setting, set by the user, to always overrides the `theme.json` setting.

justintadlock commented:
I'd argue that preexisting block settings should take precedence, regardless of what the active theme supports. This would be consistent with other attributes. For example, if I set a custom text color for a Paragraph block with a theme that has custom colors enabled, that same text color persists if I switch to a theme where custom colors are disabled.

I wouldn't break from the precedent set with other block settings, and I don't know of any other instances where this works differently.

But that's a larger problem that I wouldn't worry about specifically for this ticket (don't want to get too far into the weeds). The goal should be to display the control based on the same conditionals present in WP 6.4.

artemiomorales committed:
Fix and clarify component rendering logic

Fix issue wherein lightbox popover was rendering
erroneously when a link had been configured.

artemiomorales committed:
Reset lightbox attributes when removing link

artemiomorales committed:
Show lightbox UI if block-level override differs from default

In some cases, such as when lightbox settings already exist for
a block when global lightbox settings in theme.json change, we
should allow users to see the lightbox UI and change the settings
if they conflict with the global settings, even if the lightbox UI
is disabled globally. This prevents a block from getting stuck with
legacy lightbox settings and allows users to reset the block-level
lightbox settings if need be in these edge cases.

Note: We do not display the UI if the block-level settings exist
and match the global settings, as the block will behave as expected
in those circumstances and showing the UI in those circumstances
would likely just be confusing.

artemiomorales committed:
Handle edge case of removing existing link when lightbox is fully enabled

michalczaplinski commented:
Apologies, I didn't get to review it fully today. I ll finish up tomorrow.

michalczaplinski commented on packages/block-library/src/image/image.js:
```diff
@@ -274,6 +274,21 @@ export default function Image( {
 	}
 
+	function resetLightbox() {
+		// When deleting a link from an image while lightbox settings
+		// are enabled by default, we should disable the lightbox,
+		// otherwise the resulting UX looks like a mistake.
```
I want to make sure that I understand correctly: 

The case of image block with BOTH a link and a lightbox enabled should technically never happen. Or, at least it's not possible to get into this state using the UI. It might still happen if someone e.g. edits the attributes manually. So we have to handle this edge case somehow. Is that right?

artemiomorales replied:
This one is actually a pretty common case! If at any time, a user has the following in their `theme.json`...
```
{
	"$schema": "https://schemas.wp.org/trunk/theme.json",
	"version": 2,
	"settings": {
		"blocks": {
			"core/image": {
				"lightbox": {
					"allowEditing": true,
					"enabled": true
				}
			}
		}
	}
}
```
...and also has a link configured for the image, then one potential way to handle that UX would be to show the popup for the link, and if that link is removed, then immediately show the popup for the lightbox. However, I found that UX to be confusing.

Here I go over more in depth what the issue is. Let me know what you think, and I'm also happy to discuss more or answer any questions 😄

https://github.com/WordPress/gutenberg/assets/5360536/32065bda-d229-43ed-8655-3af4c5efffb2

michalczaplinski replied:
ok, I understand it now. Thank you for the explanation! 

I agree that this is a better UX than just "jumping" to the lightbox. 👍 

Could you add a link to this PR in this very comment in the code (line 280)? It might be helpful for someone in the future to refer to your video explanations.

artemiomorales replied:
Done 🙂

michalczaplinski reviewed (approved):
Awesome work, thank you for the exhaustive video explanation - it made reviewing much easier 🙂 

The changes seem good to me 👍 

If there is both a link and a lightbox `enabled: true` on the block we have to handle it in a slightly odd way (as I mention in [this comment](https://github.com/WordPress/gutenberg/pull/59890/files?diff=split&w=1#r1532286123)) but I think it's a very unlikely edge case and I can't think of a better way.

It would also be good to have some e2e test coverage for this logic but it's out of scope for this PR.

justintadlock reviewed (approved):
It took me a bit to work through some of the scenarios/combos, but this makes sense to me.

artemiomorales commented:
> It would also be good to have some e2e test coverage for this logic but it's out of scope for this PR.

@michalczaplinski My plan is to start working on e2e tests this week to make sure all of these scenarios are covered 👍

artemiomorales committed:
Fix focus loss preventing end-to-end test from passing

artemiomorales committed:
Add link to PR in comment

getdave commented:
This PR was triaged by release leads and contributors in WP Slack as part of the final WordPress 6.5 process. 

See https://wordpress.slack.com/archives/C02QB2JS7/p1711370484796489 (requires registration).

The criteria used was:

- Was this bug identified during the RC cycle for WP 6.5?
- Is this bug directly related to one of the blessed features in this release?
- Would this bug cause existing website’s to break/fail upon upgrade to WP 6.5?
- What is the approximate scope of the impact? Scale: (Minimal) 1 - 5 (Severe).
- Is the fix introducing any new features or changes?

Conclusion: this fix will be _not_ be included in WP 6.5 but will be punted to WP 6.5.1.

tellthemachines commented:
I just cherry-picked this PR to the add/fixes-for-651 branch to get it included in the next release: b54a426676
